### ISSUE TITLE
`needless_pass_by_ref_mut` in wrong situation with 1.73.0-nightly

### ISSUE THREAD

**Issue by ensh63** [rust-lang/rust-clippy#11222]
### Summary

There is a problem with `needless_pass_by_ref_mut` lint in 1.73.0-nightly version of clippy; it may be reported in wrong situation.

### Lint Name

needless_pass_by_ref_mut

### Reproducer

I tried this code:

```rust
use tokio::io::{ AsyncRead, AsyncReadExt };

pub async fn test_fn<I>(input: &mut I) -> Result<usize, std::io::Error>
where I: AsyncRead + Unpin,
{
    let mut buffer = vec![0u8; 32];
    input.read_exact(&mut buffer).await
}
```

I saw this happen:

```
warning: this argument is a mutable reference, but not used mutably
 --> crates/experiment/src/lib.rs:3:32
  |
3 | pub async fn test_fn<I>(input: &mut I) -> Result<usize, std::io::Error>
  |                                ^^^^^^ help: consider changing to: `&I`
  |
  = warning: changing this function will impact semver compatibility
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_ref_mut
  = note: `#[warn(clippy::needless_pass_by_ref_mut)]` on by default
```

I expected to see no warnings for this code because of the signature of `read_exec()`  method:

```rust
pub fn read_exact<'a>(&'a mut self, buf: &'a mut [u8]) -> ReadExact<'a, Self>
where
    Self: Unpin,
```

### Version

```text
rustc 1.73.0-nightly (8771282d4 2023-07-23)
binary: rustc
commit-hash: 8771282d4e7a5c4569e49d1f878fb3ba90a974d0
commit-date: 2023-07-23
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```

**Comment by Centri3** [comment]
afaict this is a duplicate of #11179, and the fix will be in nightly in a few days. If not feel free to reopen this ^^
